doc/namespace.md

```diff
@@ -0,0 +1,417 @@
+
+That means ISeq inline caches work well even with namespaces. Otherwise, it's a bug.
+
+#### Method call global cache (gccct)
+
+`rb_funcall()` C function refers to the global cc cache table (gccct), and the cache key is calculated with the current namespace.
+
+So, `rb_funcall()` calls have a performance penalty when namespace is enabled.
+
+#### Current namespace and loading namespace
+
+The current namespace is the namespace that the executing code is in. `Namespace.current` returns the current namespace object.
+
+The loading namespace is an internally managed namespace to determine the namespace to load newly required/loaded files. For example, `ns` is the loading namespace when `ns.require("foo")` is called.
+
+## Discussions
+
+#### Namespace#inspect
+
+Currently, `Namespace#inspect` returns values like `"#<Namespace:0x00000001083a5660>"`. This results in the very redundant and poorly visible classpath outside the namespace.
+
+```ruby
+# foo.rb
+class C; end
+
+# main.rb
+ns = Namespace.new
+ns.require('foo')
+
+p ns::C # "#<Namespace:0x00000001083a5660>::C"
+```
+
+And currently, if a namespace is assigned to a constant `NS1`, the classpath output will be `NS1::C`. But the namespace object can be brought to another namespace and the constant `NS1` in the namespace is something different. So the constant-based classpath for namespace is not safe basically.
+
+So we should find a better format to show namespaces. Options are:
+
+* `NS1::C` (only when this namespace is created and assigned to NS1 in the current namespace)
+* `#<Namespace:user:1083a5660>::C` (with namespace type and without preceding 0)
+* or something else
+
+#### Namespace#eval
+
+Testing namespace features needs to create files to be loaded in namespaces. It's not easy nor casual.
+
+If `Namespace` class has an instance method `#eval` to evaluate code in the namespace, it can be helpful.
+
+#### More builtin methods written in Ruby
+
+If namespace is enabled by default, builtin methods can be written in Ruby because it can't be overridden by users' monkey patches. Builtin Ruby methods can be JIT-ed, and it could bring performance reward.
+
+#### Monkey patching methods called by builtin methods
+
+Builtin methods sometimes call other builtin methods. For example, `Hash#map` calls `Hash#each` to retrieve entries to be mapped. Without namespace, Ruby users can overwrite `Hash#each` and expect the behavior change of `Hash#map` as a result.
+
+But with namespaces, `Hash#map` runs in the root namespace. Ruby users can define `Hash#each` only in user namespaces, so users cannot change `Hash#map`'s behavior in this case. To achieve it, users should override both`Hash#map` and `Hash#each` (or only `Hash#map`).
+
+It is a breaking change.
+
+It's an option to change the behavior of methods in the root namespace to refer to definitions in user namespaces. But if we do so, that means we can't proceed with "More builtin methods written in Ruby".
```

Comment:
> It's an option to change the behavior of methods in the root namespace to refer to definitions in user namespaces.

How would that work though, e.g. if multiple namespaces have different Hash#each definitions?
You would always consider the methods of the Namespace.current also in calls inside builtin methods I guess.
That then becomes a problem for method lookup inline caching.